src/coord/src/coord.rs

```diff
@@ -2291,13 +2353,44 @@ impl Coordinator {
 
     async fn sequence_peek(
         &mut self,
-        conn_id: u32,
+        session: &mut Session,
         source: MirRelationExpr,
         when: PeekWhen,
         finishing: RowSetFinishing,
         copy_to: Option<CopyFormat>,
     ) -> Result<ExecuteResponse, CoordError> {
-        let timestamp = self.determine_timestamp(&source, when)?;
+        let conn_id = session.conn_id();
+        let in_transaction = matches!(
+            session.transaction(),
+            &TransactionStatus::InTransaction(_) | &TransactionStatus::InTransactionImplicit(_)
+        );
+        // For explicit or implicit transactions that do not use AS OF, get the
+        // timestamp of the in-progress transaction or create one. If this is an AS OF
+        // query, we don't care about any possible transaction timestamp. If this is a
```

Comment:
I am afraid that mixing ```AS OF``` with non-AS-OF queries can cause issues and therefore needs to be restricted or handled better. Consider the following case:

1. A customer has a multi-statement SELECT transaction
2. They complain about queries returning stale data and is not happy
3. ```AS OF NOW()``` is suggested as a solution
4. The selects in the transaction are now inconsistent, especially if ```AS OF NOW()``` was used instead of ```AS OF CURRENT_TIMESTAMP()``` (which in turn is broken (#5903)), or only some of the statements in the transaction were  modified.

For example:

```
materialize=> begin;
BEGIN
materialize=> select mz_logical_timestamp() from t1 as of now();
 mz_logical_timestamp 
----------------------
        1614343229811
(1 row)

materialize=> select mz_logical_timestamp() from t1;
 mz_logical_timestamp 
----------------------
        1614343233896
(1 row)
```

The timestamp has advanced within a single transaction.

Possible solutions to this are:
1. Require that ```AS OF``` queries are run only outside of a transaction
2. Allow only the first query of a transaction to be an ```AS OF``` one. Keep the logical timestamp used in this query and apply it to all future queries within the same transaction.
3. Allow multiple ```AS OF``` queries in a transaction only if all the queries in a transaction are ```AS OF CURRENT_TIMESTAMP()```.

Reply:
What you describe here is exactly what I intended and I think it's ok? If you think it's incorrect or confusing to users it's pretty much just a policy decision that we can change in a later PR. I don't think it needs to be resolved now.

Reply:
I will open a separate issue about this one, because I think it is consistency violation that can actually happen in real life.